File: trees_and_graphs/path_sum.java
public boolean dfs(TreeNode node, int targetSum, int curr)
{
    if (node == null)
    {
        return false;
    }

    if (node.left == null && node.right == null)
    {
        return (curr + node.val) == targetSum;
    }

    curr += node.val;
    boolean left = dfs(node.left, targetSum, curr);
    boolean right = dfs(node.right, targetSum, curr);
    return left || right;
}

public boolean hasPathSumRecursive(TreeNode root, int targetSum)
{
    return dfs(root, targetSum, 0);
}

class Pair
{
    TreeNode node;
    int curr;
    Pair(TreeNode node, int curr)
    {
        this.node = node;
        this.curr = curr;
    }
}

public boolean hasPathSumIterative(TreeNode root, int targetSum)
{
    if (root == null)
    {
        return false;
    }

    Stack<Pair> stack = new Stack<>();
    stack.push(new Pair(root, 0));

    while (!stack.empty())
    {
        Pair pair = stack.pop();
        TreeNode node = pair.node;
        int curr = pair.curr;

        if (node.left == null
            && node.right == null
            && curr + node.val == targetSum)
        {
            return true;
        }

        curr += node.val;
        if (node.left != null)
        {
            stack.push(new Pair(node.left, curr));
        }

        if (node.right != null)
        {
            stack.push(new Pair(node.right, curr));
        }
    }

    return false;
}

File: trees_and_graphs/path_sum.py
import treenode

def hasPathSumRecursive(root: treenode.TreeNode, targetSum: int) -> bool:
    def dfs(node, curr):
        if not node:
            return False
        
        # If both children are null then the node is a leaf
        if node.left == None and node.right == None:
            return (curr + node.val) == targetSum
        
        curr += node.val
        left = dfs(node.left, curr)
        right = dfs(node.right, curr)
        return left or right
    
    return dfs(root, 0)

def hasPathSumIterative(root: treenode.TreeNode, targetSum: int) -> bool:
    if not root:
        return False
    
    stack = [(root, 0)]
    while stack:
        node, curr = stack.pop()

        # If both children are null then the node is a leaf
        if node.left == None and node.right == None:
            if (curr + node.val) == targetSum:
                return True
        
        curr += node.val
        if node.left:
            stack.append((node.left, curr))

        if node.right:
            stack.append((node.right, curr))
    
    return False